internal/controllers/machine/machine_controller_noderef.go

```diff
@@ -345,11 +356,104 @@ func (r *Reconciler) patchNode(ctx context.Context, remoteClient client.Client,
 		hasTaintChanges = taints.RemoveNodeTaint(newNode, clusterv1.NodeOutdatedRevisionTaint) || hasTaintChanges
 	}
 
-	if !hasAnnotationChanges && !hasLabelChanges && !hasTaintChanges {
+	if !hasAnnotationChanges && !hasLabelChanges && !hasTaintChanges && !propagateTaintsChanges {
+		return nil
+	}
+
+	// Use optimistic locking to avoid conflicts with other controllers.
+	mergeOptions := []client.MergeFromOption{client.MergeFromWithOptimisticLock{}}
+
+	return remoteClient.Patch(ctx, newNode, client.StrategicMergeFrom(node, mergeOptions...))
+}
+
+// propagateMachineTaintsToNode handles propagation of taints defined on a machine to a node.
+// It makes use of the annotation clusterv1.TaintsFromMachineAnnotation to track which taints are owned by the controller.
+// OnInitialization taints are only added to the node if the tracking annotation has not been set yet.
+func propagateMachineTaintsToNode(node *corev1.Node, machineTaints []clusterv1.MachineTaint) (bool, error) {
+	changed := false
+
+	// Get the value of the tracking annotation. If it is not set at all we also have to add the OnInitialization taints.
+	oldTaintsAnnotation, nodeTaintsInitialized := node.Annotations[clusterv1.TaintsFromMachineAnnotation]
+
+	// ownedTaints contains all Always taints that the controller is owning.
+	ownedTaints := unmarshalMachineTaintsAnnotation(oldTaintsAnnotation)
+
+	// newOwnedTaints will contain all Always taints from the current machine's spec.
+	newOwnedTaints := sets.New[string]()
+	onInitializationTaints := sets.New[string]()
+
+	for _, taint := range machineTaints {
+		// Collect Always and OnInitialization taints to identify taints to delete.
+		// Separating Always taints so the tracking annotation can be updated accordingly.
+		switch taint.Propagation {
+		case clusterv1.MachineTaintPropagationAlways:
+			newOwnedTaints.Insert(fmt.Sprintf("%s:%s", taint.Key, taint.Effect))
+		case clusterv1.MachineTaintPropagationOnInitialization:
+			onInitializationTaints.Insert(fmt.Sprintf("%s:%s", taint.Key, taint.Effect))
+		}
+
+		// Only add OnInitialization taints if the tracking annotation has not been set yet.
+		if taint.Propagation == clusterv1.MachineTaintPropagationOnInitialization && nodeTaintsInitialized {
+			continue
+		}
+
+		// Ensure the taint is set on the node and has the correct value.
+		if changedTaints := taints.EnsureNodeTaint(node, convertMachineTaintToCoreV1Taint(taint)); changedTaints {
```

Comment:
This does not update the value. Let's also extend the test coverage

Reply:
Great catch, fixed and also added some more coverage for other cases too.